shared-libs/cht-datasource/src/libs/core.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * A value that could be <code>null</code>.
+ */
+export type Nullable<T> = T | null;
+
+/**
+ * An array that is guaranteed to have at least one element.
+ */
+export type NonEmptyArray<T> = [T, ...T[]];
+
+type DataPrimitive = string | number | boolean | Date | null | undefined;
+interface DataArray extends Readonly<(DataPrimitive | DataArray | DataObject)[]> { }
+export interface DataObject extends Readonly<Record<string, DataPrimitive | DataArray | DataObject>> { }
+
+export const isString = (value: unknown): value is string => {
+  return typeof value === 'string';
+};
+
+export const isRecord = (value: unknown): value is Record<string, unknown> => {
+  return value !== null && typeof value === 'object';
+};
+
+export const hasField = (value: Record<string, unknown>, field: { name: string, type: string }): boolean => {
+  const valueField = value[field.name];
+  return typeof valueField === field.type;
+};
+
+export const hasFields = (
+  value: Record<string, unknown>,
+  fields: NonEmptyArray<{ name: string, type: string }>
+): boolean => fields.every(field => hasField(value, field));
```

Comment:
Not super excited to have logic like this.  I think it is important to be able to check the shape of data provided at the edges of the library (e.g. getting passed in as an argument or coming back from Couch).  I think a "schema validation" library (e.g. https://zod.dev/) basically does this (and a lot more).  We should consider pulling in something like that, but perhaps not for this MVP... :thinking:  :shrug:

Reply:
I had great experiences with zod in the past. We use [joi](https://joi.dev/) in cht-conf which is pretty similar but for typescript codebases zod has the upper hand. You're right that we don't need it right now, it's an implementation detail that can easily be switched out at a later point :+1:

Reply:
Have to say, it seems pretty rare to hear a developer say they "had great experiences" with a particular library!  We all have a million libs that we hate to use, but ones that were great?  Those are precious!  :laughing: